tremor-cli/src/job.rs

```diff
@@ -0,0 +1,192 @@
+                            stdout_tx.send(resp).map_err(|e| e.into())
+                        })
+                    }));
+
+                    let stderr_thread = Some(thread::spawn(move || -> Result<()> {
+                        // Redirect target process stderr
+                        readlines_until_eof(stderr, |resp| {
+                            stderr_tx.send(resp).map_err(|e| e.into())
+                        })
+                    }));
+
+                    Ok(Self {
+                        process: target_cmd,
+                        stdout_thread,
+                        stderr_thread,
+                        stdout_receiver: stdout_rx,
+                        stderr_receiver: stderr_rx,
+                    })
+                } else {
+                    Err("Unable to create error stream from target process".into())
+                }
+            } else {
+                Err("Unable to create output stream from target process".into())
+            }
+        } else {
+            Err("Unable to create target process".into())
+        }
+    }
+
+    pub fn wait_with_output(&mut self) -> Result<std::process::ExitStatus> {
+        Ok(self.process.wait()?)
+    }
+
+    pub(crate) fn tail(&mut self, stdout_path: &str, stderr_path: &str) -> Result<()> {
+        let mut tailout = File::create(stdout_path)?;
+        let mut tailerr = File::create(stderr_path)?;
+
+        self.wait_with_output()?;
+
+        loop {
+            match self.stdout_receiver.try_recv() {
+                Ok(line) => {
+                    tailout.write(line.as_bytes())?;
+                }
+                Err(TryRecvError::Empty) => (),
+                Err(TryRecvError::Disconnected) => {
+                    break;
+                }
+            }
+        }
+
+        loop {
+            match self.stderr_receiver.try_recv() {
+                Ok(line) => {
+                    tailerr.write(line.as_bytes())?;
+                }
+                Err(TryRecvError::Empty) => (),
+                Err(TryRecvError::Disconnected) => {
+                    break;
+                }
+            }
+        }
+
+        tailout.sync_all()?;
+        tailerr.sync_all()?;
+        if let Err(_) = self.process.kill() {
+            // Do nothing, it already exited
+        }
+        Ok(())
+    }
+}
+
+impl Drop for TargetProcess {
+    fn drop(&mut self) {
+        if let Some(handle) = self.stdout_thread.take() {
+            handle
+                .join()
+                .unwrap_or_else(|_| {
+                    Ok(()) // FIXME error handling
+                })
+                .ok();
+        }
+
+        if let Some(handle) = self.stderr_thread.take() {
+            handle.join().unwrap_or_else(|_| Ok(())).ok(); // FIXME error handling
+        }
+
+        match self.process.wait() {
+            Ok(status) => if !status.success() {},
+            Err(_) => {
+                () // FIXME error handling
```

Comment:
Rather strangely this clippy error on rust latest stable doesn't fire on my dev machine